providers-sdk/v1/plugin/interface.go

```diff
@@ -32,6 +33,7 @@ type ProviderCallback interface {
 type ProviderPlugin interface {
 	ParseCLI(req *ParseCLIReq) (*ParseCLIRes, error)
 	Connect(req *ConnectReq, callback ProviderCallback) (*ConnectRes, error)
+	Disconnect(req *emptypb.Empty) error
```

Comment:
Currently, Disconnect is called short before cnspec exits. So, I'd assume none of the connections is needed any longer.

Do you think of other use cases for this func? Who would know which connection to shut down? My current understanding is, that we handle all connections inside a plugin. Shouldn't we then also shut down all these connections inside a plugin?

A lot of questions, I know. I implemented this for the use cases I could think of. Perhaps there are other ones I missed. That's why I asked so many questions.

Reply:
perhaps in a scenario like: 
open up aws connection, open up fs connection to scan filesystem, scan filesystem, close fs connection, do a couple more things, close aws connection

in that case we would want to close the fs connection before we're totally done with the aws connection, i think

Reply:
I personally cannot think of a use case, i only based this on what the `ConnectRes` looks like. I think it's fine for now to leave it like this, perhaps we could replace the google's `Empty` struct with our own that we can then extend in the future, if that's needed

Reply:
Semantics first: Disconnect being the opposite of Connect, I'd love for it to behave in a similar way. Where connect opens a connection, it would be great for disconnect to also only target one connection.

Thank you for sharing the use-case in https://github.com/mondoohq/cnquery/issues/1628 @czunker , this helped a lot! In that case, I'd rather we add a Shutdown call that is used to terminate all remaining elements in the provider plugin. 

I think Disconnect for specific connections is also a great idea. For now, it looks like the Shutdown of the provider is all we need?

(A note on this down the line: The use-case where we have a ton on assets discovered that create a ton of connections is one where something like disconnect for specific connections will be useful. We don't optimize for this case just yet, but after v9 is finished I think it's something we can start to look into, especially when memory issues start to rise. There is also a case for starting individual providers for connections, which will slow down some of the exchange, but rapidly accelerate memory cleanup on shutdown ;) )

One more thing on Shutdown: Please make sure to include a sane timeout for cleanup actions, so we don't get stuck forever. It's fine to do this in a follow-up, doesn't have to be right away...

Reply:
Thanks al for the feedback.

I renamed the func to Shutdown, added a timeout, and incorporated some other feedback. Please have another look.

I tried it with the gcp snapshot PR and it worked there.

About the sane timeout:
"sane" is a bit difficult here. I opted for 2 minutes. The only timeout I found in the code is for VMWare, which is 5 minutes. I didn't want to go too low, because in the GCP example we are calling cloud APIs. So, hopefully 2 minutes is fine.

It might be a solution to add it as an option to providers in one of the upcoming iterations. Or use a CLI flag to set it.